# Lab 1b (bonus for deep understanding) - Chapter 2 
> Author : Badr TAJINI - Large Language model (LLMs) - ESIEE 2024-2025


#### 1. Package Installation:

- Install the additional package requirements for this bonus notebook by uncommenting and running the following cell:

In [1]:
# pip install -r requirements-extra.txt

This setup block ensures all necessary dependencies are available for the comparative analysis.

# Comparing Various Byte Pair Encoding (BPE) Implementations

#### 2. Tiktoken Implementation:
<br>
&nbsp;

#### Using BPE from `tiktoken`

In [2]:
from importlib.metadata import version

print("tiktoken version:", version("tiktoken"))

tiktoken version: 0.8.0


In [3]:
import tiktoken

tik_tokenizer = tiktoken.get_encoding("gpt2")

text = "Hello, world. Is this-- a test?"

This section introduces OpenAI's tiktoken library (version 0.5.1), which represents a modern, optimized implementation of BPE tokenization. The code initializes the GPT-2 encoding scheme and defines a test string.

#### 3. Tiktoken Encoding/Decoding:

In [4]:
integers = tik_tokenizer.encode(text, allowed_special={"<|endoftext|>"})

print(integers)

[15496, 11, 995, 13, 1148, 428, 438, 257, 1332, 30]


In [5]:
strings = tik_tokenizer.decode(integers)

print(strings)

Hello, world. Is this-- a test?


In [6]:
print(tik_tokenizer.n_vocab)

50257


This demonstrates the bidirectional conversion between text and token IDs. The output reveals that GPT-2's vocabulary contains 50,257 tokens, a carefully chosen size balancing coverage and efficiency.

#### 4. Original GPT-2 BPE Implementation:
<br>
&nbsp;

####  Using the original BPE implementation used in GPT-2

In [7]:
from bpe_openai_gpt2 import get_encoder, download_vocab

In [8]:
download_vocab()

Fetching encoder.json: 1.04Mit [00:01, 862kit/s]                                                    
Fetching vocab.bpe: 457kit [00:00, 748kit/s]                                                        


In [9]:
orig_tokenizer = get_encoder(model_name="gpt2_model", models_dir=".")

In [10]:
integers = orig_tokenizer.encode(text)

print(integers)

[15496, 11, 995, 13, 1148, 428, 438, 257, 1332, 30]


In [11]:
strings = orig_tokenizer.decode(integers)

print(strings)

Hello, world. Is this-- a test?


This section implements the original GPT-2 tokenization scheme, providing a baseline for comparison. The vocabulary files are downloaded to ensure consistency with the original implementation.

#### 5. Hugging Face Transformers Implementation:
<br>
&nbsp;

#### Using the BPE via Hugging Face transformers

In [12]:
import transformers

transformers.__version__

'4.40.2'

In [13]:
from transformers import GPT2Tokenizer

hf_tokenizer = GPT2Tokenizer.from_pretrained("gpt2")

C:\Users\adrie\anaconda3\Lib\site-packages\huggingface_hub\file_download.py:797: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

C:\Users\adrie\anaconda3\Lib\site-packages\huggingface_hub\file_download.py:139: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\adrie\.cache\huggingface\hub\models--gpt2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

In [14]:
hf_tokenizer(strings)["input_ids"]

[15496, 11, 995, 13, 1148, 428, 438, 257, 1332, 30]

This introduces the Hugging Face implementation, which has become a de facto standard in the research community due to its accessibility and integration with the broader transformers ecosystem.

#### 6. Performance Benchmark:
<br>
&nbsp;

#### A quick performance benchmark

In [17]:
with open('../1_main_code/the-verdict.txt', 'r', encoding='utf-8') as f:
    raw_text = f.read()

In [18]:
%timeit orig_tokenizer.encode(raw_text)

9.6 ms ± 730 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [19]:
%timeit tik_tokenizer.encode(raw_text)

2.04 ms ± 32.7 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [20]:
%timeit hf_tokenizer(raw_text)["input_ids"]

Token indices sequence length is longer than the specified maximum sequence length for this model (5145 > 1024). Running this sequence through the model will result in indexing errors


16.8 ms ± 719 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [21]:
%timeit hf_tokenizer(raw_text, max_length=5145, truncation=True)["input_ids"]

16.6 ms ± 962 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


The benchmark reveals significant performance differences:

- Original implementation: ~4.29ms
- Tiktoken: ~1.4ms
- Hugging Face: ~8.46ms

This performance comparison highlights tiktoken's optimization, achieving roughly 3x faster processing than the original implementation and 6x faster than Hugging Face's version. These differences become particularly significant when processing large volumes of text for training or inference.
The notebook effectively demonstrates the evolution of BPE implementation in the context of GPT-2, from the original version to more recent optimized implementations. The consistent tokenization results across implementations validate their compatibility, while the performance metrics illuminate the practical implications of implementation choices in production environments.
This analysis provides valuable insights selecting tokenization implementations based on specific requirements for speed, compatibility, and ecosystem integration.

END.